### PR TITLE
Deleting host from etcd.

### PR DESCRIPTION
**Pull request by jeremykross**
@nicktate @normanjoyner 

-- there may be more to do here, but this seems to shore up deleteHost quite a bit.

### CODE REVIEW

File: lib/kubernetes-api.js [hunk 1]
```diff
@@ -487,6 +489,7 @@ class KubernetesApi extends ApiImplementation {
                 app_desc.count = desired_count;
             }
 
+
```
Comment (nicktate) [inline lib/kubernetes-api.js]: remove extra newline

File: lib/kubernetes-api.js [hunk 2]
```diff
@@ -202,6 +203,7 @@ class KubernetesApi extends ApiImplementation {
             services: (cb) => {
                 return request(`${this.k8s_api_url}/api/v1/namespaces/default/services`,
                     (err, res, body) => {
+
```
Comment (nicktate) [inline lib/kubernetes-api.js]: remove extra newline

File: lib/kubernetes-api.js [hunk 3]
```diff
@@ -192,6 +192,7 @@ class KubernetesApi extends ApiImplementation {
             replication_controllers: (cb) => {
                 return request(`${this.k8s_api_url}/api/v1/namespaces/default/replicationcontrollers`,
                     (err, res, body) => {
+
```
Comment (nicktate) [inline lib/kubernetes-api.js]: remove extra newline

File: lib/kubernetes-api.js [hunk 4]
```diff
@@ -784,28 +787,106 @@ class KubernetesApi extends ApiImplementation {
...
-        return request({
-            uri: `${this.k8s_api_url}/api/v1/nodes/${host_id}`,
-            method: 'DELETE'
-        }, (err, res, body) => {
-            if(err || res.statusCode !== 200) {
-                return cb(err || body);
+        async.parallel({
+            hosts: (cb) => {
+                //Get the hosts
+                return this.getHosts((err, hosts) => {
```
Comment (nicktate) [inline lib/kubernetes-api.js]: `return this.getHosts(cb)`

File: lib/kubernetes-api.js [hunk 5]
```diff
@@ -784,28 +787,106 @@ class KubernetesApi extends ApiImplementation {
...
+
+                //Get etcd members.
+                request({
+                    uri: `${this.etcd_api_url}/v2/members`,
+                    method: 'GET',
+                    json: true
+                }, (err, res) => {
+
+                    if(err) {
+                        logger.error(`Error determing etcd host for k8s deletion on ${host_id}: ${err}.`);
```
Comment (nicktate) [inline lib/kubernetes-api.js]: spell: determining

File: lib/kubernetes-api.js [hunk 6]
```diff
@@ -833,6 +915,7 @@ class KubernetesApi extends ApiImplementation {
                     method: 'GET',
                     uri: `${this.k8s_api_url}/version`
                 }, (err, res, body) => {
+
```
Comment (nicktate) [inline lib/kubernetes-api.js]: remove extra newline

File: lib/kubernetes-api.js [hunk 7]
```diff
@@ -816,6 +897,7 @@ class KubernetesApi extends ApiImplementation {
             services: (cb) => {
                 return request(`${this.k8s_api_url}/api/v1/services`,
                     (err, res, body) => {
+
```
Comment (nicktate) [inline lib/kubernetes-api.js]: remove extra newline

File: lib/kubernetes-api.js [hunk 8]
```diff
@@ -784,28 +787,106 @@ class KubernetesApi extends ApiImplementation {
...
-    //Look here!
-    //Use os module to get info about master!
+        //Look here!
+        //Use os module to get info about master!
     getHosts(cb) {
         return async.parallel({
             nodes: (cb) => {
                 return request(`${this.k8s_api_url}/api/v1/nodes`,
                     (err, res, body) => {
+
```
Comment (nicktate) [inline lib/kubernetes-api.js]: remove extra newline

File: lib/kubernetes-api.js [hunk 9]
```diff
@@ -784,28 +787,106 @@ class KubernetesApi extends ApiImplementation {
...
 
-            return cb();
         });
+
     }
 
-    //Look here!
-    //Use os module to get info about master!
+        //Look here!
+        //Use os module to get info about master!
```
Comment (nicktate) [inline lib/kubernetes-api.js]: remove extra space

File: lib/kubernetes-api.js [hunk 10]
```diff
@@ -784,28 +787,106 @@ class KubernetesApi extends ApiImplementation {
...
+            });
 
-            return cb();
         });
+
     }
 
-    //Look here!
-    //Use os module to get info about master!
+        //Look here!
```
Comment (nicktate) [inline lib/kubernetes-api.js]: remove extra space

File: lib/kubernetes-api.js [hunk 11]
```diff
@@ -784,28 +787,106 @@ class KubernetesApi extends ApiImplementation {
...
+
+            const {hosts, etcd_members} = result;
+
+            async.waterfall([
+                (cb) => {
+                    const host = _.get(hosts, host_id);
+                    const host_url = _.get(host, ['address', 'public']);
+
+                    const etcd_member_id = _.flow(
+                        _.partial(_.filter, _, (m) => {
```
Comment (nicktate) [inline lib/kubernetes-api.js]: spell out variables for readability

`m -> member`

File: lib/kubernetes-api.js [hunk 12]
```diff
@@ -784,28 +787,106 @@ class KubernetesApi extends ApiImplementation {
...
+                logger.error(`Couldn't fetch requisite data in deleteHost(): ${err}`);
+                return cb(err);
+            }
+
+            const {hosts, etcd_members} = result;
+
+            async.waterfall([
+                (cb) => {
+                    const host = _.get(hosts, host_id);
+                    const host_url = _.get(host, ['address', 'public']);
```
Comment (nicktate) [inline lib/kubernetes-api.js]: `host_public_ip`

File: lib/kubernetes-api.js [hunk 13]
```diff
@@ -784,28 +787,106 @@ class KubernetesApi extends ApiImplementation {
...
+
+                    const etcd_member_id = _.flow(
+                        _.partial(_.filter, _, (m) => {
+                            const client_urls = _.get(m, 'clientURLs', []);
+                            return _.some(client_urls, (c_url) => _.includes(c_url, host_url));
```
Comment (nicktate) [inline lib/kubernetes-api.js]: `c_url` -> `client_url`

Comment (nicktate) [inline lib/kubernetes-api.js]: host_public_ip

Comment (nicktate) [inline lib/kubernetes-api.js]: c_url should probably be c_ip anyway?

File: lib/kubernetes-api.js [hunk 14]
```diff
@@ -784,28 +787,106 @@ class KubernetesApi extends ApiImplementation {
...
+
+                        cb();
+                    });
+                },
+                (cb) =>  {
+                    return request({
+                        uri: `${this.k8s_api_url}/api/v1/nodes/${host_id}`,
+                        method: 'DELETE'
+                    }, (err, res) => {
+
```
Comment (nicktate) [inline lib/kubernetes-api.js]: remove extra newline

File: lib/kubernetes-api.js [hunk 15]
```diff
@@ -784,28 +787,106 @@ class KubernetesApi extends ApiImplementation {
...
+                        if(err || !Math.floor(res.statusCode / 200) === 1) {
+                            logger.error(`Error deleting k8s host ${host_id}.`);
+                            return cb(err || res);
+                        }
+
+                        cb(null, res);
+                    });
+                }
+            ], (err, k8s_result) => {
+
```
Comment (nicktate) [inline lib/kubernetes-api.js]: remove extra newline

File: lib/kubernetes-api.js [hunk 16]
```diff
@@ -784,28 +787,106 @@ class KubernetesApi extends ApiImplementation {
...
+                        uri: `${this.k8s_api_url}/api/v1/nodes/${host_id}`,
+                        method: 'DELETE'
+                    }, (err, res) => {
+
+                        if(err || !Math.floor(res.statusCode / 200) === 1) {
+                            logger.error(`Error deleting k8s host ${host_id}.`);
+                            return cb(err || res);
+                        }
+
+                        cb(null, res);
```
Comment (nicktate) [inline lib/kubernetes-api.js]: return

File: lib/kubernetes-api.js [hunk 17]
```diff
@@ -784,28 +787,106 @@ class KubernetesApi extends ApiImplementation {
...
+                    request({
+                        uri: `${this.etcd_api_url}/v2/members/${etcd_member_id}`,
+                        method: 'DELETE'
+                    }, (err, res) => {
+                        if(err || !Math.floor(res.statusCode / 200) === 1) {
+                            logger.error(`Error deleting etcd host ${etcd_member_id}: ${err}`);
+                            return cb(err || res);
+                        }
+
+                        cb();
```
Comment (nicktate) [inline lib/kubernetes-api.js]: return

File: lib/kubernetes-api.js [hunk 18]
```diff
@@ -784,28 +787,106 @@ class KubernetesApi extends ApiImplementation {
...
+                //Get etcd members.
+                request({
+                    uri: `${this.etcd_api_url}/v2/members`,
+                    method: 'GET',
+                    json: true
+                }, (err, res) => {
+
+                    if(err) {
+                        logger.error(`Error determing etcd host for k8s deletion on ${host_id}: ${err}.`);
+                        cb(err);
```
Comment (nicktate) [inline lib/kubernetes-api.js]: return

File: lib/kubernetes-api.js [hunk 19]
```diff
@@ -784,28 +787,106 @@ class KubernetesApi extends ApiImplementation {
...
+            },
+            etcd_members: (cb) => {
+
+                //Get etcd members.
+                request({
+                    uri: `${this.etcd_api_url}/v2/members`,
+                    method: 'GET',
+                    json: true
+                }, (err, res) => {
+
```
Comment (nicktate) [inline lib/kubernetes-api.js]: remove extra newline

File: lib/kubernetes-api.js [hunk 20]
```diff
@@ -784,28 +787,106 @@ class KubernetesApi extends ApiImplementation {
...
-                return cb(err || body);
+        async.parallel({
+            hosts: (cb) => {
+                //Get the hosts
+                return this.getHosts((err, hosts) => {
+                    cb(err, hosts);
+                });
+            },
+            etcd_members: (cb) => {
+
```
Comment (nicktate) [inline lib/kubernetes-api.js]: remove extra newline

File: lib/kubernetes-api.js [hunk 21]
```diff
@@ -784,28 +787,106 @@ class KubernetesApi extends ApiImplementation {
...
+                }
+            ], (err, k8s_result) => {
+
+                if(err) {
+                    logger.error(`Caught an error during final callback during deleteHost: ${err}.`);
+                    return cb(err);
+                }
+
+                this.enforceAllConstraints();
+                return cb(null, k8s_result);
```
Comment (nicktate) [inline lib/kubernetes-api.js]: delete host should not return any data

File: lib/kubernetes-api.js (hunk 13 above)
Comment (nicktate) [inline lib/kubernetes-api.js]: I could be off on the url -> ip thought here, just verify you are correctly describing the data

Comment (jeremykross) [inline lib/kubernetes-api.js]: Well, it's coming from a k8s key called cluentURLs.

**Comment by jeremykross** [comment]
@nicktate updated.

File: lib/kubernetes-api.js [hunk 22]
```diff
@@ -784,23 +784,93 @@ class KubernetesApi extends ApiImplementation {
...
-        }, (err, res, body) => {
-            if(err || res.statusCode !== 200) {
-                return cb(err || body);
+        async.parallel({
+            hosts: (cb) => {
+                //Get the hosts
+                return this.getHosts(cb);
+            },
+            etcd_members: (cb) => {
+
```
Comment (normanjoyner) [inline lib/kubernetes-api.js]: remove whitespace

File: lib/kubernetes-api.js [hunk 23]
```diff
@@ -784,23 +784,93 @@ class KubernetesApi extends ApiImplementation {
...
     deleteHost(host_id, cb) {
-        return request({
-            uri: `${this.k8s_api_url}/api/v1/nodes/${host_id}`,
-            method: 'DELETE'
-        }, (err, res, body) => {
-            if(err || res.statusCode !== 200) {
-                return cb(err || body);
+        async.parallel({
+            hosts: (cb) => {
+                //Get the hosts
```
Comment (normanjoyner) [inline lib/kubernetes-api.js]: can we move these comments outside of function as they are useful in describing what each step of the parallel is accomplishing

File: lib/kubernetes-api.js [hunk 24]
```diff
@@ -784,23 +784,93 @@ class KubernetesApi extends ApiImplementation {
...
+        }, (err, result) => {
 
-            // immediately trigger constraint update since host count has changed
-            this.enforceAllConstraints();
+            if(err) {
+                logger.error(`Couldn't fetch requisite data in deleteHost(): ${err}`);
+                return cb(err);
+            }
+
+            const {hosts, etcd_members} = result;
```
Comment (normanjoyner) [inline lib/kubernetes-api.js]: can we do `{ hosts, etcd_members }` for readability?

File: lib/kubernetes-api.js [hunk 25]
```diff
@@ -784,23 +784,93 @@ class KubernetesApi extends ApiImplementation {
...
+
+            async.waterfall([
```
Comment (normanjoyner) [inline lib/kubernetes-api.js]: comment waterfall steps

File: lib/kubernetes-api.js [hunk 26]
```diff
@@ -784,23 +784,93 @@ class KubernetesApi extends ApiImplementation {
...
+
+                    if(!etcd_member_id) {
+                        return cb(Error('Couldn\'t find associated etcd member id.'));
+                    }
+
+                    request({
+                        uri: `${this.etcd_api_url}/v2/members/${etcd_member_id}`,
+                        method: 'DELETE'
+                    }, (err, res) => {
+                        if(err || !Math.floor(res.statusCode / 200) === 1) {
```
Comment (normanjoyner) [inline lib/kubernetes-api.js]: `!==` when checking all of these response status codes

File: lib/kubernetes-api.js [hunk 27]
```diff
@@ -784,23 +784,93 @@ class KubernetesApi extends ApiImplementation {
...
+                            logger.error(`Error deleting k8s host ${host_id}.`);
+                            return cb(err || res);
+                        }
+
+                        return cb(null, res);
+                    });
+                }
+            ], (err, k8s_result) => {
```
Comment (normanjoyner) [inline lib/kubernetes-api.js]: having `k8s_result` here will probably trigger lint warning as it's unused

File: lib/kubernetes-api.js [hunk 28]
```diff
@@ -784,23 +784,93 @@ class KubernetesApi extends ApiImplementation {
...
+                logger.error(`Couldn't fetch requisite data in deleteHost(): ${err}`);
+                return cb(err);
+            }
+
+            const {hosts, etcd_members} = result;
+
+            async.waterfall([
+                (cb) => {
+                    const host = _.get(hosts, host_id);
+                    const host_url = _.get(host, ['host_public_ip']);
```
Comment (normanjoyner) [inline lib/kubernetes-api.js]: `const host_url = host && host.host_public_ip;`

File: lib/kubernetes-api.js [hunk 29]
```diff
@@ -784,23 +784,93 @@ class KubernetesApi extends ApiImplementation {
...
+            if(err) {
+                logger.error(`Couldn't fetch requisite data in deleteHost(): ${err}`);
+                return cb(err);
+            }
+
+            const {hosts, etcd_members} = result;
+
+            async.waterfall([
+                (cb) => {
+                    const host = _.get(hosts, host_id);
```
Comment (normanjoyner) [inline lib/kubernetes-api.js]: not sure how we want to tackle this style-wise, but if we prefer dot notation for fetching top level keys, we should use bracket notation here

File: lib/kubernetes-api.js (hunk 25 above)
Comment (normanjoyner) [inline lib/kubernetes-api.js]: if we only care about the result of the final async function, we can use `async.series` here

**Comment by jeremykross** [comment]
You guys are much more sensitive to aesthetics than I am - probably a good thing - I'll be more mindful of it in the future.

**Pull request merged**